### PR TITLE
support cstring in `case`

### LINKED ISSUE #350: Accept cstrings for case selector in JS
This code:

```nim
let foo = cstring"abc"
case $foo
of "abc": echo "yes"
else: echo "no"
```

outputs:

```js
var foo_1872002 = "abc";
switch (toJSStr(cstrToNimstr(foo_1872002))) {
case "abc":
  rawEcho(makeNimstrLit("yes"));
  break;
default: 
  rawEcho(makeNimstrLit("no"));
  break;
}
```

The `toJSStr` call is always emitted, and if it's a conversion from cstring to string you get a pointless `toJSStr(cstrToNimstr(foo))`. If the selector could be `cstring` we would be able to remove the `toJSStr` call and not have to convert with `cstrToNimstr`.

This is only relevant to JS so it would only be allowed on it, if this intrudes on consistency between backends (which I really doubt will be much of an issue to anyone) and it's easy to support cstring for the custom string case in C backend, then that could also be done, but if it's not worth it we could just deal with the inconsistency.

Someone might want to mention this even though it's unrelated, it's also very easy for the matcher values in the `of` branches to accept cstring values as they are the same in the VM, but this is irrelevant to the backend and would not affect performance.

### PR DESCRIPTION
**Pull request by metagn**
closes https://github.com/nim-lang/RFCs/issues/350

Again a bit sloppy, but should work.

Can add experimental switch or whatever if desired.

### CODE REVIEW

**Comment by juancarlospaco** [comment]
Does it works when the `cstring` is `nil` ?.

**Comment by metagn** [comment]
I thought it would be complicated to make it work but it's not

File: compiler/ccgstmts.nim [hunk 1]
```diff
@@ -843,8 +843,13 @@ proc genCaseStringBranch(p: BProc, b: PNode, e: TLoc, labl: TLabel,
...
     initLocExpr(p, b[i], x)
-    assert(b[i].kind in {nkStrLit..nkTripleStrLit})
-    var j = int(hashString(p.config, b[i].strVal) and high(branches))
+    var j: int
+    case b[i].kind
+    of nkStrLit..nkTripleStrLit:
+      j = int(hashString(p.config, b[i].strVal) and high(branches))
+    of nkNilLit: j = 0
+    else:
+      assert false, "invalid string case branch node kind"
```
Comment (juancarlospaco) [inline compiler/ccgstmts.nim:852]:
```suggestion
      doAssert false, "invalid string case branch node kind"
```

Comment (metagn) [reply compiler/ccgstmts.nim:852]: Was previously `assert` + there are other surrounding `assert`s. No special need to force doAssert here

**Comment by Araq** [comment]
This is awesome! Since this will be in version 2.0 it doesn't need an experimental switch. It does need a changelog entry though and documentation updates.

**Comment by metagn** [comment]
Done

**Pull request merged**